### ISSUE TITLE
Path NotFoundError when downloading bundles

### ISSUE THREAD

**Issue by AndrewJGaut** [codalab/codalab-worksheets#4529]
We get NotFoundError: Path '' in bundle <uuid> not found for some bundles. This occurs when you try to download the bundle (or even `get_target_info` for the bundle).

**NOTE**: This is blocking completion of migration to Azure Blob storage, which is urgent because it greatly impacts funding.

The full stack trace Jiani got in the migration script for one such bundle is included below:
```
2023-09-11 02:15:33,968 [migration] In get_bundle_info, 0x3dcfb5bf6ee648b8b634b19699cbc1cf /home/azureuser/codalab-worksheets/var/codalab/home/partitions/default/bundles/0x3dcfb5bf6ee648b8b634b19699cbc1cf 0x3dcfb5bf6ee648b8b634b19699cbc1cf:
2023-09-11 02:15:33,969 [migration] Error: Path ‘’ in bundle 0x3dcfb5bf6ee648b8b634b19699cbc1cf not found
Traceback (most recent call last):
  File “codalab/migration.py”, line 359, in <module>
    bundle_info = migration.get_bundle_info(bundle_uuid, bundle_location)
  File “codalab/migration.py”, line 121, in get_bundle_info
    raise e
  File “codalab/migration.py”, line 117, in get_bundle_info
    info = download_util.get_target_info(bundle_location, target, depth=0)
  File “/opt/codalab-worksheets/codalab/worker/download_util.py”, line 105, in get_target_info
    “Path ‘{}’ in bundle {} not found”.format(target.subpath, target.bundle_uuid)
codalab.worker.download_util.PathException: Path ‘’ in bundle 0x3dcfb5bf6ee648b8b634b19699cbc1cf not found
```

**Comment by AndrewJGaut** [comment]
Looks like this is occurring for almost every bundle in dev that's older than a certain date! Very odd...

**Comment by AndrewJGaut** [comment]
I believe I've localized it: looks like we get this error when downloading ANY bundle with ID less than or equal to `1319295` in the dev database. For all bundles with ID greater than that, downloading works. 

That last bundle that doesn't work has UUID `0xee06050835ab41ecb7ec859c65f105b5` and was uploaded at `2023-07-16 16:57:37`, and the first bundle that does work has UUID `0xc4ca82339da843b680ef9a04f1e5161c` and was uploaded at `2023-07-17 01:56:28`.

Looks like we merged a breaking change at some point around mid-July. Investigating further.

**Comment by AndrewJGaut** [comment]
That was right around when we made [this](https://github.com/codalab/codalab-worksheets/commit/2992b374b813055a23214b31a063bfe9d19c7db9) change, FYI...

**Comment by AndrewJGaut** [comment]
Looking more at it, though, the error happens [here](https://github.com/codalab/codalab-worksheets/blob/v1.7.1/codalab/worker/download_util.py#L102C10-L107). It's clear, then, that for non-linked bundles (which is most of them), the bundles just don't exist on disk.

Moreover, this doesn't happen on prod.

I now suspect that we just cleared out the dev disks without adjusting the database on July 16 or July 17, and that's causing the issue. Will try and confirm this now.

**Comment by AndrewJGaut** [comment]
Ah, OK, I was wrong before: it isn't that all bundles before a certain ID can't be downloaded; it appears that it's just that a BUNCH (the vast majority) of bundles just aren't present in the bundle store on disk on the dev server. You can see some with ID less than the one I posted as being the cutoff (e.g. `0x202062d8b1464d6eac426e103920bd8e`, with ID `1308999`, can be downloaded).

You can see all the bundles that actually exist in the bundle store in the `/home/azureuser/codalab-worksheets/var/codalab/home/partitions/default/bundles` directory in the `codalab_rest-server_1` container on `vm-clws-dev-0`. I'm not sure why the other bundles got wiped, but this is the cause of the issue: The path isn't found for some bundles because that bundle literally doesn't exist on disk!

**Comment by AndrewJGaut** [comment]
As a side note: we may want to improve the error message given there. (e.g. rather than 'Path not found', which isn't very descriptive, we could use 'Bundle does not exist in bundle store' or something like this when the path provided id `''`, since that always means the bundle just does not exist)

**Comment by AndrewJGaut** [comment]
Now, looking at statistics, on dev, we have:
**809 bundles on disk in bundle store**
**111,486 on disk in database**


Showing work here:
- Finding amount in bundle store
```
root@4f19506a6fe1:/home/azureuser/codalab-worksheets/var/codalab/home/partitions/default/bundles# ls -lth | wc -l
809
```
- Finding amount in database
```
mysql> select count(*) from bundle where storage_type='disk';
+----------+
| count(*) |
+----------+
|   111486 |
+----------+
1 row in set (0.03 sec)
```

**Comment by AndrewJGaut** [comment]
Repeating the analysis on prod: Looks like it's still an issue, but not nearly as bad as on dev. We have:
**171,129 on disk in bundle store**
**181,774 on disk in database**
So about 94% of the on-disk bundles are actually present.


Showing work:
- Finding amount in bundle store
```
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls
codalab0  codalab1  codalab10  codalab2  codalab3  codalab4  codalab5  codalab6  codalab7  codalab8  codalab9
oot@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab0/bundles | wc -l
12335
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab1/bundles | wc -l
11219
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab2/bundles | wc -l
14849
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab3/bundles | wc -l
13070
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab4/bundles | wc -l
12245
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab5/bundles | wc -l
10778
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab6/bundles | wc -l
8627
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab7/bundles | wc -l
10553
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab8/bundles | wc -l
37447
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab9/bundles | wc -l
30122
root@af003715d4b4:/home/azureuser/codalab-worksheets/var/codalab/home/partitions# ls -lth codalab10/bundles | wc -l
9884
```
`12335 + 11219 + 14849 + 13070 + 12245 + 10778 + 8627 + 10553 + 37447 + 30122 + 9884 = 171129`
- Finding amount in database:
```
mysql> select count(*) from bundle where storage_type='disk';
+----------+
| count(*) |
+----------+
|   181774 |
+----------+
1 row in set (0.04 sec)
```

**Comment by AndrewJGaut** [comment]
The TL;DR:

- On dev, a lot (about **99%!**) of bundles that exist in the database and are said to be stored on-disk in the database don't actually exist on disk.

- When you try to download those bundles (or get their target info), an error is thrown.

- The migration script is having this error thrown, since it needs to get their target info.

- That error itself is NOT indicative of an issue with the migration script in its current version.

- The discrepancy between on-disk bundles in the bundle store and in the database exists in prod, too, but is not as bad (about 6% of bundles in the database that are said to be on-disk don't exist on disk).

- I'm not yet sure WHY the discrepancy exists.

**Comment by AndrewJGaut** [comment]
Action items:
- Further analysis on non-present bundles on prod:
  - How old are they? 
  - Do they have any special properties we can identify?
  - Do we see anything in the prod error logs that would indicate a problem there? Perhaps database not being updated upon bundle deletion?
- Cleaning up dev
  - Delete the non-present bundles from the database.
- Proceed with migration
  - This is not an issue with the script I believe; we should keep testing on dev.
  - If the issue gets worse on dev as we test, though, we will now know that the script is the cause since we have hard numbers for the bundle discrepancy
- Make error more descriptive.
  - See [here](https://github.com/codalab/codalab-worksheets/issues/4529#issuecomment-1716900634)

**Comment by AndrewJGaut** [comment]
[difference.txt](https://github.com/codalab/codalab-worksheets/files/12668735/difference.txt)
Attached are all the UUIDs of bundles in the database but not present on disk.

To get this, I queried the prod database like so:
```
docker exec -it mysqlserver_mysql_1 mysql -u bundles_user -p -D codalab_bundles -e "select uuid from bundle where storage_type='disk'" -s -N | tee /tmp/test-output.txt;
```
and I used a small script to get all bundles on disk on the prod server like so:
```
partition=<partition-path>

num_partitions=10
for ((i=0;i<=10;i+=1));
do
  dirpath="${partition}${i}"
  echo $dirpath;
  docker exec -ti codalab_rest-server_1 ls $dirpath/bundles -l | awk '{print $9 }' >> /tmp/bundles_on_disk.txt
done
```

Then, I wrote a small python script to get the difference between the two sets of UUIDs.

Now, we can do some analysis on the missing bundle UUIDs.

**Comment by AndrewJGaut** [comment]
OK, I compiled some data using a script that ran `cl info` on all those bundles. (One thing to note: You can get `info` on 95 bundles per command max, but this GREATLY speeds up the collection of results over getting `info` for bundles one by one.)

I'm attaching that data as well as some analysis I did on about 14,000 out of the about 18,000 missing bundles in prod (the script will collect info for all bundles, but it hasn't quite finished yet). I couldn't export to PDF on my machine, so I just include the notebook as well as some screenshots.
[information.csv](https://github.com/codalab/codalab-worksheets/files/12670118/information.csv)
<img width="1026" alt="Screen Shot 2023-09-20 at 12 32 12 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/b6ef41cf-ed4b-49a3-a702-60806d806ac9">
<img width="1009" alt="Screen Shot 2023-09-20 at 12 32 18 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/850087e8-7226-4413-8476-ca2e26b4808c">
<img width="1012" alt="Screen Shot 2023-09-20 at 12 32 23 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/e06ae89e-d19d-447b-9046-9b76711bb93f">
<img width="1020" alt="Screen Shot 2023-09-20 at 12 32 28 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/e4ffac33-b672-4ce2-b53a-2db8cbf3cb3a">
<img width="1025" alt="Screen Shot 2023-09-20 at 12 32 33 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/c1c65c0a-304b-408b-a1d7-6618f2085c84">
<img width="1023" alt="Screen Shot 2023-09-20 at 12 32 50 AM" src="https://github.com/codalab/codalab-worksheets/assets/35617203/f9a15e59-60d9-43ff-a060-c4e802736116">







So far, I haven't found anything particularly special about these bundles... but perhaps someone else will look at the data and have better insights.

**Comment by AndrewJGaut** [comment]
OK, so, after discussing in meeting, this is probably all due to `cl rm -d` commands being run. On dev, there likely was another cause as well, but it was probably due to some testing that went awry several years ago. Therefore, I'm closing the issue.
